src/treelearner/serial_tree_learner.cpp

```diff
@@ -892,4 +936,56 @@ void SerialTreeLearner::RenewTreeOutput(Tree* tree, const ObjectiveFunction* obj
   }
 }
 
+// this function computes the best split for a given leaf and a given feature
+void SerialTreeLearner::ComputeBestSplitForFeature(
+    double sum_gradient, double sum_hessian, data_size_t num_data,
+    int feature_index, FeatureHistogram *histogram_array_,
+    std::vector<SplitInfo> &bests, int leaf_index, int depth, const int tid,
+    int real_fidx, LearnerState &learner_state, bool update) {
+
+  // if this is not a subtree stemming from a monotone split, then no constraint
+  // apply
+  if (learner_state.tree->leaf_is_in_monotone_subtree(leaf_index)) {
+    if (!learner_state.config_->monotone_precise_mode) {
+      learner_state.current_constraints.Set(learner_state.constraints_per_leaf_[leaf_index],
+                              tid);
+    }
+  }
+
+#ifdef DEBUG
+  learner_state.current_constraints.CheckCoherenceWithLeafOutput(
+      learner_state.tree->LeafOutput(leaf_index), tid, kEpsilon)
+#endif
+      SplitInfo new_split;
+
+  SplittingConstraints *constraints;
+  if (learner_state.config_->monotone_constraints.empty()) {
+    constraints = nullptr;
+  } else {
+    constraints = &learner_state.current_constraints[tid];
+  }
+  histogram_array_[feature_index].FindBestThreshold(
+      sum_gradient, sum_hessian, num_data, &new_split, constraints);
+
+  if (learner_state.tree->leaf_is_in_monotone_subtree(leaf_index)) {
+    learner_state.current_constraints.InitializeConstraints(tid);
+  }
+
+  new_split.feature = real_fidx;
+  if (learner_state.cegb_ != nullptr) {
+      new_split.gain -= learner_state.cegb_->DetlaGain(feature_index, real_fidx, leaf_index, num_data, new_split);
+  }
+
+
+  if (new_split.monotone_type != 0) {
+    double penalty = LeafConstraints::ComputeMonotoneSplitGainPenalty(
+        depth, learner_state.config_->monotone_penalty);
+    new_split.gain *= penalty;
+  }
+
+  if (new_split > bests[tid]) {
+    bests[tid] = new_split;
+  }
+}
+
```

Comment:
I think we should avoid the extra calls in serial_tree_learner and feature_histogram as possible.

Reply:
I am sorry, but I don't understand what functions in particular you are referring to. I think the code, right now, when there are no constraints, is working in a very similar way to they way it used to work. But I am also happy to make the changes you request if you can be more specific about this.